### PR TITLE
Issue 5751: Added dmScript::CheckGOInstance

### LINKED ISSUE #5751: Add dmScript::CheckGOInstance
We need a function to get the current game object from the `dmScript` namespace.

### PR DESCRIPTION
**Pull request by JCash**

### CODE REVIEW

File: engine/gamesys/src/gamesys/gamesys_script.cpp [hunk 1]
```diff
@@ -35,12 +35,64 @@
...
+            if (scene != 0) {
+                instance = (dmGameObject::HInstance)dmGameSystem::GuiGetUserDataCallback(scene);
+            }
+        }
+        // No instance for render scripts, ignored
+        if (instance == 0) {
+            luaL_error(L, "no instance could be found in the current script environment");
+        }
+        return instance;
+    }
```
Comment (JCash) [inline engine/gamesys/src/gamesys/gamesys_script.cpp:60]: For functions that doesn't take a game object reference as input

File: engine/gamesys/src/gamesys/gamesys_script.cpp [hunk 2]
```diff
@@ -35,12 +35,64 @@
...
+            instance = dmGameObject::GetInstanceFromIdentifier(collection, receiver.m_Path);
+            if (!instance)
+            {
+                luaL_error(L, "Instance %s not found", lua_tostring(L, instance_arg));
+                return 0; // Actually never reached
+            }
+        }
+        return instance;
+    }
+}
```
Comment (JCash) [inline engine/gamesys/src/gamesys/gamesys_script.cpp:93]: For functions that use an optional reference argument

File: engine/gamesys/src/gamesys/gamesys_script.cpp [hunk 3]
```diff
@@ -87,18 +139,7 @@ namespace dmGameSystem
...
-            if (scene != 0) {
-                instance = (dmGameObject::HInstance)GuiGetUserDataCallback(scene);
-            }
-        }
-        // No instance for render scripts, ignored
-        if (instance == 0) {
-            luaL_error(L, "no instance could be found in the current script environment");
-        }
-        return instance;
+        return dmScript::CheckGOInstance(L);
```
Comment (JCash) [inline engine/gamesys/src/gamesys/gamesys_script.cpp:142]: I didn't want to update all the current users of this (albeit internal) function.

File: engine/gamesys/src/wscript [hunk 4]
```diff
@@ -10,6 +11,8 @@ def build(bld):
     bld.add_subdirs('gamesys')
     bld.install_files('${PREFIX}/lib/python', 'waf_gamesys.py')
 
+    dmsdk_add_files(bld, '${PREFIX}/sdk/include/dmsdk', 'dmsdk')
```
Comment (JCash) [inline engine/gamesys/src/wscript:14]: Added dmsdk support for gamesys library

File: engine/gamesys/src/dmsdk/gamesys/script.h [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+     *     lua_pushnumber(L, position.getX());
+     *     lua_pushnumber(L, position.getY());
+     *     lua_pushnumber(L, position.getZ());
+     *     return 3;
+     * }
+     * ```
+     */
+    dmGameObject::HInstance CheckGOInstance(lua_State* L, int index);
+
+    /*# buffer ownership
```
Comment (JCash) [inline engine/gamesys/src/dmsdk/gamesys/script.h:67]: The following code was previously in the _lower_ level library include file `script/script.h` which is imho not a good solution (include file in one lib, source code in another)

File: engine/gamesys/src/dmsdk/script.h [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+#ifndef DMSDK_SCRIPT_H
+#define DMSDK_SCRIPT_H
+
+#include <dmsdk/script/script.h>
+#include <dmsdk/gamesys/script.h>
+
+#endif // DMSDK_SCRIPT_H
```
Comment (JCash) [inline engine/gamesys/src/dmsdk/script.h:19]: Created this new overarching script.h that references the difference implementations.

File: engine/sdk/src/dmsdk/sdk.h [hunk 7]
```diff
@@ -15,7 +15,7 @@
 
 #include <dmsdk/extension/extension.h>
 #include <dmsdk/engine/extension.h>
-#include <dmsdk/script/script.h>
+#include <dmsdk/script.h> // references both script/script.h and gamesys/script.h
```
Comment (JCash) [inline engine/sdk/src/dmsdk/sdk.h:18]: This should make the change invisible to our users

File: engine/gamesys/src/dmsdk/gamesys/script.h [hunk 8]
```diff
@@ -0,0 +1,168 @@
+// Copyright 2020 The Defold Foundation
```
Comment (Jhonnyg) [inline engine/gamesys/src/dmsdk/gamesys/script.h:1]: Should we update this stub now?

Comment (Jhonnyg) [reply engine/gamesys/src/dmsdk/gamesys/script.h:1]: Ah nevermind if this file was moved

**Review by Jhonnyg — APPROVED** [review]
Looks good!

File: engine/gamesys/src/dmsdk/gamesys/script.h (hunk 8 above)
Comment (JCash) [reply engine/gamesys/src/dmsdk/gamesys/script.h:1]: Yes, we probably should do a separate PR with a license update.

**Pull request merged**